Sources/tart/Fetcher.swift

```diff
@@ -0,0 +1,46 @@
+import Foundation
+import AsyncAlgorithms
+
+class Fetcher: NSObject, URLSessionTaskDelegate, URLSessionDelegate, URLSessionDataDelegate {
+    let responseCh = AsyncThrowingChannel<URLResponse, Error>()
+    let dataCh = AsyncThrowingChannel<Data, Error>()
+
+    func fetch(_ request: URLRequest) async throws -> (AsyncThrowingChannel<Data, Error>, URLResponse) {
+        let session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
+        let task = session.dataTask(with: request)
+        task.resume()
+
+        // Wait for the response and only then return
+        var iter = responseCh.makeAsyncIterator()
+        let response = try await iter.next()!
+
+        return (dataCh, response)
+    }
+
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse) async -> URLSession.ResponseDisposition {
+        await responseCh.send(response)
+
+        return .allow
+    }
+
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
+        let sema = DispatchSemaphore(value: 0)
```

Comment:
Why there is a need for a semaphore?

Reply:
In order to wait for a `Task` to terminate, otherwise we will get results out-of-order. We also can't use `Task.result` because it's `async`, but we're in a `sync` context.

See also https://stackoverflow.com/a/42484670/9316533 and https://stackoverflow.com/a/47041695/9316533.